packages/jest-matcher-utils/src/deepCyclicCopyReplaceable.ts

```diff
@@ -59,7 +59,9 @@ function deepCyclicCopyObject<T>(object: T, cycles: WeakMap<any, any>): T {
       descriptor.value = deepCyclicCopyReplaceable(descriptor.value, cycles);
     }
 
-    if (!('set' in descriptor)) {
+    if ('set' in descriptor) {
```

Comment:
Why can't all of this just be:

```js
const descriptor = descriptors[key];
descriptor = {
  configurable: true,
  enumerable: true,
  writeable: true,
  value: object[key]
}
```

?

Reply:
By the way I don't think `if (typeof descriptor.value !== 'undefined') {` above is right given that `undefined` itself *is* a valid value. The correct check I think is `descriptor.hasOwnProperty('value')`. But regardless, I don't yet see why we can't do it simpler like I wrote above.

Reply:
In theory we probably could, but it makes the assumption that either the getter always returns the same thing, or that the property is only retrieved from the deep copied object once (because in the case of a getter it snapshots the first return value instead of preserving the getter).
The current solution is a more 'accurate' deep copy, although not sure if it matters, because I doubt we support getters with changing return values anyway (the deep equals call and this diff call could already get different values).

Reply:
yeah, that seems to work as well! pushed

Reply:
>In theory we probably could, but it makes the assumption that either the getter always returns the same thing, or that the property is only retrieved from the deep copied object once (because in the case of a getter it snapshots the first return value instead of preserving the getter).

I think in case of printing we _want_ to call the getter as few times as possible, right? Since otherwise the thing we're printing might be different from the thing we were comparing.

That might actually be a good stress test for this whole thing. Have a getter that returns different things on every call. Verify that the comparison shows the diff for the correct result.

Reply:
Yes, I agree that once for every deep property would be ideal. But I think that goal is far away as we already have other get calls happening in a deep equals assertion. Having only one would require the combined deepEqual-prettyPrint-diff implementation (aka the 'proper solution') that we never got around to implementing.
I was viewing this more from a 'deep copy' perspective, in which case preserving getters would be more accurate. This way, it's more of a 'snapshot for diff' method.

Reply:
I don't feel too strongly about this though and think it would be fine to land this way, especially since it's not our standard `deepCyclicCopy` but `deepCyclicCopyReplaceable`.

Reply:
>But I think that goal is far away as we already have other get calls happening in a deep equals assertion.

That seems not great. Is that new behavior in Jest 25, or always been that way?

Reply:
Afaik it has always been like that, because `expect`, `pretty-format`, and `jest-diff` are separate things as opposed to one unified algorithm that generates a 'diff data structure'. `expect` takes the values and checks whether they match (first access to properties). If not, it feeds them into `jest-diff`, which `pretty-format`s both (second access to properties) and then just does a string diff on the two result.
This (as opposed to a unified algorithm for all three tasks) has caused quite some problems esp. regarding asymmetric matchers in the past. I'd love to rewrite the whole thing, but it's huge :sweat_smile:

Reply:
To put the code we're talking about into context, it was introduced recently as a workaround for the asymmetric matchers issues (not the proper unified algorithm fix we'd like) and runs when expect has detected a mismatch, but before the diffing, as a sort of preprocessing.

Reply:
Makes sense.